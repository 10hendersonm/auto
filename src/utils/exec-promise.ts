import * as childProcess from 'child_process';
import { promisify } from 'util';

const exec = promisify(childProcess.exec);

export default async function execPromise(cmd: string) {
  const { stdout, stderr } = await exec(cmd, { maxBuffer: 1024 * 1000 });

  if (stderr) {
    console.log(stderr);
  }

  return stdout.trim();
}
