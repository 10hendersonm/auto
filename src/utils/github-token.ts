import fs from 'fs';
import path from 'path';
import registryUrl from 'registry-url';
import { promisify } from 'util';
import settingsUrl from './settings-url';

const readFile = promisify(fs.readFile);
const registry = registryUrl();
// tslint:disable-next-line
const normalizedRegistry = registry.replace('http:', '').replace('https:', '');

export default async function getGitHubToken(
  apiUrl = 'https://api.github.com'
): Promise<string> {
  if (process.env.GH_TOKEN) {
    return process.env.GH_TOKEN;
  }

  const helpText = `Try setting an access token up ${settingsUrl(apiUrl)}`;

  if (!process.env.HOME) {
    throw new Error(`Can't find the GH_TOKEN and no HOME defined. ${helpText}`);
  }

  const rcLocation = path.resolve(process.env.HOME, '.npmrc');

  if (fs.existsSync(rcLocation)) {
    const rc = await readFile(rcLocation, 'utf8');

    const regex = new RegExp(`${normalizedRegistry}:_authToken=\(\\w+\)`);
    const token = rc.match(regex);
    if (token && token[1]) {
      return token[1];
    }

    throw new Error(`No token in the .npmrc. ${helpText}`);
  }

  throw new Error(`Can't find a GitHub token to use. ${helpText}`);
}
