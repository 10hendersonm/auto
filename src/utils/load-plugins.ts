import * as path from 'path';
import { AutoRelease } from '../main';
import NPMPlugin from '../plugins/npm';

export type IPluginConstructor = new () => IPlugin;

export interface IPlugin {
  name: string;
  apply(auto: AutoRelease): void;
}

type SupportedPlugin = 'npm';
const plugins = new Map<SupportedPlugin, IPluginConstructor>([
  ['npm', NPMPlugin]
]);

function isSupported(key: SupportedPlugin | string): key is SupportedPlugin {
  return !!plugins.get(key as SupportedPlugin);
}

export default function loadPlugin(
  pluginPath: SupportedPlugin | string
): IPluginConstructor | undefined {
  if (isSupported(pluginPath)) {
    return plugins.get(pluginPath);
  }

  try {
    return require(pluginPath);
  } catch (error) {
    return require(path.join(process.cwd(), pluginPath));
  }
}
