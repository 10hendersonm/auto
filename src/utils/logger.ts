import signale, { Signale } from 'signale';

export interface ILogger {
  log: signale.Signale<signale.DefaultMethods>;
  verbose: signale.Signale<signale.DefaultMethods>;
  veryVerbose: signale.Signale<signale.DefaultMethods>;
}

export function dummyLog(): ILogger {
  return {
    log: new Signale({ disabled: true }),
    verbose: new Signale({ disabled: true }),
    veryVerbose: new Signale({ disabled: true })
  };
}

export type LogLevel = 'verbose' | 'veryVerbose' | undefined;

export default function createLog(mode: LogLevel): ILogger {
  return {
    log: new Signale(),
    verbose: new Signale({
      disabled: mode !== 'verbose' && mode !== 'veryVerbose'
    }),
    veryVerbose: new Signale({
      disabled: mode !== 'veryVerbose'
    })
  };
}
