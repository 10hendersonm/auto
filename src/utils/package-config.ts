import fs from 'fs';
import parseGithubUrl from 'parse-github-url';
import { promisify } from 'util';

export interface IRepoConfig {
  owner: string;
  repo: string;
  version: string;
}

interface IPackageConfig {
  version: string;
  repository?:
    | {
        url: string;
      }
    | 'string';
}

const readFile = promisify(fs.readFile);

const getPackageConfig = async (): Promise<IPackageConfig> => {
  const pkgConfig = await readFile('./package.json', 'utf-8');
  return JSON.parse(pkgConfig);
};

export default async function getConfigFromPackageJson(): Promise<IRepoConfig> {
  const { repository, version } = await getPackageConfig();

  if (!repository) {
    throw new Error('Cannot read repo info from package.json');
  }

  const { owner, name } =
    parseGithubUrl(
      typeof repository === 'string' ? repository : repository.url
    ) || ({} as any);

  if (!owner || !name) {
    throw new Error(
      'Cannot read owner and package name from Github URL in package.json'
    );
  }

  return {
    repo: name,
    owner,
    version
  };
}
