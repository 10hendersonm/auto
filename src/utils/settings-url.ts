import * as url from 'url';

export default function settingsUrl(apiUrl: string) {
  let { host } = url.parse(apiUrl);

  if (host && host.includes('api.')) {
    host = host.replace('api.', '');
  }

  return `https://${host}/settings/tokens/new`;
}
