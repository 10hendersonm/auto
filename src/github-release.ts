import GHub from '@octokit/rest';
import * as fs from 'fs';
import { ICommit } from 'gitlog';
import { inc, ReleaseType } from 'semver';
import signale from 'signale';
import { promisify } from 'util';

import GitHub, { IPRInfo } from './git';
import generateReleaseNotes, {
  IExtendedCommit,
  normalizeCommits
} from './log-parse';
import SEMVER, { calculateSemVerBump, IVersionLabels } from './semver';
import exec from './utils/exec-promise';
import { dummyLog } from './utils/logger';
import postToSlack from './utils/slack';

export type VersionLabel =
  | SEMVER.major
  | SEMVER.minor
  | SEMVER.patch
  | 'skip-release'
  | 'release'
  | 'prerelease';

export const defaultLabels = new Map<VersionLabel, string>();
defaultLabels.set(SEMVER.major, 'major');
defaultLabels.set(SEMVER.minor, 'minor');
defaultLabels.set(SEMVER.patch, 'patch');
defaultLabels.set('skip-release', 'skip-release');
defaultLabels.set('release', 'release');
defaultLabels.set('prerelease', 'prerelease');

export const defaultChangelogTitles = {
  major: 'üí•  Breaking Change',
  minor: 'üöÄ  Enhancement',
  patch: 'üêõ  Bug Fix',
  internal: 'üè†  Internal',
  documentation: 'üìù  Documentation'
};

export const defaultLabelsDescriptions = new Map<string, string>();
defaultLabelsDescriptions.set(SEMVER.major, 'create a major release');
defaultLabelsDescriptions.set(SEMVER.minor, 'create a minor release');
defaultLabelsDescriptions.set(SEMVER.patch, 'create a patch release');
defaultLabelsDescriptions.set('skip-release', 'do not create a release');
defaultLabelsDescriptions.set(
  'release',
  'publish a release when this pr is merged'
);
defaultLabelsDescriptions.set('prerelease', 'create pre release');
defaultLabelsDescriptions.set(
  'internal',
  'changes are internal to the project'
);
defaultLabelsDescriptions.set(
  'documentation',
  'changes only effect documentation'
);

export interface ILogger {
  log: signale.Signale<signale.DefaultMethods>;
  verbose: signale.Signale<signale.DefaultMethods>;
  veryVerbose: signale.Signale<signale.DefaultMethods>;
}

export interface IGitHubReleaseOptions {
  platform?: 'npm';
  labels?: {
    [label: string]: string;
  };
  logger: ILogger;
  jira?: string;
  slack?: string;
  githubApi?: string;
  name?: string;
  email?: string;
  skipReleaseLabels: string[];
  onlyPublishWithReleaseLabel?: boolean;
  'no-version-prefix'?: boolean;
  changelogTitles?: {
    [label: string]: string;
  };
}

export interface IOptionalGitHubOptions {
  owner?: string;
  repo?: string;
  baseUrl?: string;
  token?: string;
}

const readFile = promisify(fs.readFile);
const writeFile = promisify(fs.writeFile);

export default class GitHubRelease {
  private readonly github: GitHub;
  private readonly releaseOptions: IGitHubReleaseOptions;
  private readonly userLabels: IVersionLabels;
  private readonly changelogTitles: { [label: string]: string };

  private readonly githubApi: string;

  constructor(
    options: IOptionalGitHubOptions,
    releaseOptions: IGitHubReleaseOptions = {
      logger: dummyLog(),
      skipReleaseLabels: []
    }
  ) {
    this.releaseOptions = releaseOptions;
    this.githubApi = releaseOptions.githubApi || 'https://api.github.com';
    this.changelogTitles = releaseOptions.changelogTitles || {};
    this.userLabels = new Map(Object.entries(releaseOptions.labels || {}) as [
      VersionLabel,
      string
    ][]);

    if (!options.owner || !options.repo || !options.token) {
      throw new Error('Must set owner, repo, and GitHub token.');
    }

    this.releaseOptions.logger.verbose.info(
      'Options contain repo information.'
    );

    const args = {
      owner: options.owner,
      repo: options.repo,
      logger: this.releaseOptions.logger,
      ...options
    };

    if (releaseOptions && this.githubApi) {
      args.baseUrl = this.githubApi;
    }

    // So that --verbose can be used on public CIs
    const tokenlessArgs = {
      ...args,
      token: args.token
        ? `[Token starting with ${args.token.substring(0, 4)}]`
        : undefined
    };

    this.releaseOptions.logger.verbose.info(
      'Initializing GitHub API with:\n',
      tokenlessArgs
    );
    this.github = new GitHub(args);
  }

  public async generateReleaseNotes(
    from: string,
    to = 'HEAD'
  ): Promise<string> {
    const commits = await this.getCommits(from, to);
    const project = await this.github.getProject();

    await Promise.all(
      commits.map(async commit => {
        commit.packages = await this.github.changedPackages(commit.hash);
      })
    );

    return generateReleaseNotes(commits, {
      owner: this.github.options.owner,
      repo: this.github.options.repo,
      baseUrl: project.data.html_url,
      jira: this.releaseOptions.jira,
      logger: this.releaseOptions.logger,
      changelogTitles: {
        ...defaultChangelogTitles,
        ...this.changelogTitles
      }
    });
  }

  public async addToChangelog(
    releaseNotes: string,
    lastRelease: string,
    currentVersion: string,
    message = 'Update CHANGELOG.md [skip ci]'
  ) {
    this.releaseOptions.logger.verbose.info('Adding new changes to changelog.');

    let version;

    if (lastRelease.match(/\d+\.\d+\.\d+/)) {
      version = await this.calcNextVersion(lastRelease);
    } else {
      const bump = await this.getSemverBump(lastRelease);
      version = inc(currentVersion, bump as ReleaseType);
    }

    this.releaseOptions.logger.verbose.info(
      'Calculated next version to be:',
      version
    );

    const date = new Date().toDateString();
    const prefixed =
      this.releaseOptions['no-version-prefix'] ||
      (version && version.startsWith('v'))
        ? version
        : `v${version}`;

    let newChangelog = `# ${prefixed} (${date})\n\n${releaseNotes}`;

    if (fs.existsSync('CHANGELOG.md')) {
      this.releaseOptions.logger.verbose.info(
        'Old changelog exists, prepending changes.'
      );
      const oldChangelog = await readFile('CHANGELOG.md', 'utf8');
      newChangelog = `${newChangelog}\n\n---\n\n${oldChangelog}`;
    }

    await writeFile('CHANGELOG.md', newChangelog);
    this.releaseOptions.logger.verbose.info(
      'Wrote new changelog to filesystem.'
    );

    await exec('git add CHANGELOG.md');
    await exec(`git commit -m '${message}' --no-verify`);
    this.releaseOptions.logger.verbose.info('Commited new changelog.');
  }

  public async getCommits(
    from: string,
    to = 'HEAD'
  ): Promise<IExtendedCommit[]> {
    this.releaseOptions.logger.verbose.info(
      `Getting commits from ${from} to ${to}`
    );

    const gitlog = await this.github.getGitLog(from, to);

    this.releaseOptions.logger.veryVerbose.info('Got gitlog:\n', gitlog);

    const commits = await this.addLabelsToCommits(gitlog);

    this.releaseOptions.logger.veryVerbose.info(
      'Added labels to commits:\n',
      commits
    );

    await Promise.all(
      commits.map(async commit => {
        let resolvedAuthors = [];

        if (commit.pullRequest) {
          const prCommits = await this.github.getCommitsForPR(
            Number(commit.pullRequest.number)
          );

          if (!prCommits) {
            return;
          }

          resolvedAuthors = await Promise.all(
            prCommits.map(async prCommit => {
              if (prCommit && prCommit.author) {
                return this.github.getUserByUsername(prCommit.author.login);
              }
            })
          );
        } else if (commit.authorEmail) {
          const author = await this.github.getUserByEmail(commit.authorEmail);
          resolvedAuthors.push(author);
        }

        commit.authors = resolvedAuthors.map(author => ({
          ...author,
          username: author ? author.login : undefined
        }));

        commit.authors.map(author => {
          this.releaseOptions.logger.veryVerbose.info(
            `Found author: ${author.username}`
          );
        });
      })
    );

    return commits;
  }

  public async publish(releaseNotes: string, tag: string) {
    return this.github.publish(releaseNotes, tag);
  }

  public async getLabels(pr: number) {
    return this.github.getLabels(pr);
  }

  public async createStatus(prInfo: IPRInfo) {
    return this.github.createStatus(prInfo);
  }

  public async getSha() {
    return this.github.getSha();
  }

  public async getLatestRelease(): Promise<string> {
    return this.github.getLatestRelease();
  }

  public async getPullRequest(pr: number) {
    return this.github.getPullRequest(pr);
  }

  public async createComment(message: string, pr: number, context = 'default') {
    return this.github.createComment(message, pr, context);
  }

  public async getPullRequests(options?: Partial<GHub.PullsListParams>) {
    return this.github.getPullRequests(options);
  }

  public async addLabelsToProject(labels: Map<string, string>) {
    const oldLabels = await this.github.getProjectLabels();
    const labelsToCreate = [...labels.entries()].filter(
      ([versionLabel, customLabel]) => {
        if (oldLabels && oldLabels.includes(customLabel)) {
          return;
        }

        if (
          versionLabel === 'release' &&
          !this.releaseOptions.onlyPublishWithReleaseLabel
        ) {
          return;
        }

        if (
          versionLabel === 'skip-release' &&
          this.releaseOptions.onlyPublishWithReleaseLabel
        ) {
          return;
        }

        return true;
      }
    );

    await Promise.all(
      labelsToCreate.map(async ([versionLabel, customLabel]) => {
        await this.github.createLabel(versionLabel, customLabel);
      })
    );

    const repoMetadata = await this.github.getRepoMetadata();

    const justLabelNames = labelsToCreate.map(([name]) => name);
    if (justLabelNames.length > 0) {
      this.releaseOptions.logger.log.log(
        `Created labels: ${justLabelNames.join(', ')}`
      );
    } else {
      this.releaseOptions.logger.log.log(
        'No labels were created, they must have already been present on your project.'
      );
    }
    this.releaseOptions.logger.log.log(
      `\nYou can see these, and more at ${repoMetadata.html_url}/labels`
    );
  }

  public async getSemverBump(from: string, to = 'HEAD'): Promise<SEMVER> {
    const commits = await this.getCommits(from, to);
    const labels = commits.map(commit => commit.labels);
    const {
      onlyPublishWithReleaseLabel,
      skipReleaseLabels
    } = this.releaseOptions;
    const options = { onlyPublishWithReleaseLabel, skipReleaseLabels };
    const versionLabels = new Map([...defaultLabels, ...this.userLabels]);

    this.releaseOptions.logger.verbose.info(
      'Calculating SEMVER bump using:\n',
      {
        labels,
        versionLabels,
        options
      }
    );

    const result = calculateSemVerBump(labels, versionLabels, options);

    this.releaseOptions.logger.verbose.success(
      'Calculated SEMVER bump:',
      result
    );

    return result;
  }

  public async postToSlack(releaseNotes: string, tag: string) {
    if (!this.releaseOptions.slack) {
      throw new Error('Slack url must be set to post a message to slack.');
    }

    const project = await this.github.getProject();

    this.releaseOptions.logger.verbose.info('Posting release notes to slack.');

    await postToSlack(releaseNotes, {
      tag,
      owner: this.github.options.owner,
      repo: this.github.options.repo,
      baseUrl: project.data.html_url,
      slackUrl: this.releaseOptions.slack
    });

    this.releaseOptions.logger.verbose.info('Posted release notes to slack.');
  }

  public async calcNextVersion(lastTag: string) {
    const bump = await this.getSemverBump(lastTag);
    return inc(lastTag, bump as ReleaseType);
  }

  private async addLabelsToCommits(commits: ICommit[]) {
    if (!commits) {
      return [];
    }

    const eCommits = normalizeCommits(commits);

    await Promise.all(
      eCommits.map(async commit => {
        if (!commit.pullRequest) {
          commit.labels = [];
        } else {
          commit.labels = await this.getLabels(
            parseInt(commit.pullRequest.number, 10)
          );
        }
      })
    );

    return eCommits;
  }
}
