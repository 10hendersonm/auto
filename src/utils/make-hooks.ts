import { AsyncSeriesBailHook, AsyncSeriesHook, SyncHook } from 'tapable';
import { IGitHubReleaseHooks } from '../github-release';
import { IChangelogHooks } from '../log-parse';
import { IAutoHooks } from '../main';

export const makeHooks = (): IAutoHooks => ({
  beforeRun: new SyncHook(['config']),
  beforeShipIt: new SyncHook([]),
  onCreateGitHubRelease: new SyncHook(['config']),
  onCreateChangelog: new SyncHook(['gitHubReleaseConfig']),
  getAuthor: new AsyncSeriesBailHook([]),
  getPreviousVersion: new AsyncSeriesBailHook(['prefixRelease']),
  getRepository: new AsyncSeriesBailHook([]),
  publish: new AsyncSeriesHook(['version'])
});

export const makeGitHubReleaseHooks = (): IGitHubReleaseHooks => ({
  onCreateChangelog: new SyncHook(['changelog'])
});

export const makeChangelogHooks = (): IChangelogHooks => ({
  renderChangelogLine: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogTitle: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogAuthor: new AsyncSeriesBailHook([
    'author',
    'commit',
    'options'
  ]),
  renderChangelogAuthorLine: new AsyncSeriesBailHook(['author', 'user'])
});
