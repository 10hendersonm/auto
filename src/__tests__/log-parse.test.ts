import { defaultLabels } from '../github-release';
import generateReleaseNotes, {
  createUserLink,
  filterServiceAccounts,
  normalizeCommits,
  parseJira,
  parsePR,
  parseSquashPR
} from '../log-parse';
import { dummyLog } from '../utils/logger';
import makeCommitFromMsg from './make-commit-from-msg';

const logger = dummyLog();

describe('createUserLink', () => {
  test('should ', () => {
    expect(
      createUserLink(
        {
          name: 'none',
          email: undefined,
          username: 'invalid-email-address'
        },
        {
          hash: '1',
          labels: [],
          pullRequest: {
            number: '22'
          },
          authorName: 'none',
          authorEmail: 'default@email.com',
          authors: [
            {
              name: 'none',
              email: undefined
            }
          ],
          subject: ''
        },
        {
          owner: '',
          repo: '',
          baseUrl: 'https://github.custom.com/',
          changelogTitles: {},
          versionLabels: defaultLabels
        }
      )
    ).toBe(undefined);
  });

  test('should find email', () => {
    expect(
      createUserLink(
        {
          name: 'none',
          email: undefined
        },
        {
          hash: '1',
          labels: [],
          pullRequest: {
            number: '22'
          },
          authorName: 'none',
          authorEmail: 'default@email.com',
          authors: [
            {
              name: 'none',
              email: undefined
            }
          ],
          subject: ''
        },
        {
          owner: '',
          repo: '',
          baseUrl: 'https://github.custom.com/',
          changelogTitles: {},
          versionLabels: defaultLabels
        }
      )
    ).toBe('default@email.com');
  });
});

describe('parsePR', () => {
  test('should do nothing without merge commit', () => {
    const commit = makeCommitFromMsg('Not a merge');
    expect(parsePR(commit)).toEqual(commit);
  });

  test('should parse number base and comment', () => {
    const commit = makeCommitFromMsg(
      'Merge pull request #1234 from Another PR\nComments about the PR'
    );

    expect(parsePR(commit)).toEqual({
      ...commit,
      pullRequest: {
        number: '1234',
        base: 'Another PR'
      },
      subject: 'Comments about the PR'
    });
  });
});

describe('parseSquashPR', () => {
  test('should do nothing squash merge commit', () => {
    const commit = makeCommitFromMsg('Not a squash');
    expect(parseSquashPR(commit)).toEqual(commit);
  });

  test('should parse PR number and comment', () => {
    const commit = makeCommitFromMsg('Some Message (#1234)');
    expect(parseSquashPR(commit)).toEqual({
      ...commit,
      pullRequest: {
        number: '1234'
      },
      subject: 'Some Message'
    });
  });
});

describe('filterServiceAccounts', () => {
  test('should not filter the commit', () => {
    const commit = makeCommitFromMsg('foo');
    expect(filterServiceAccounts(commit)).toEqual(commit);
  });

  test('should filter the commit', () => {
    const commit = {
      ...makeCommitFromMsg('foo'),
      authorName: 'pdbf',
      authorEmail: ''
    };

    expect(filterServiceAccounts(commit)).toBeUndefined();
  });
});

describe('jira', () => {
  test('no story', () => {
    const commit = {
      ...makeCommitFromMsg('Add log')
    };

    expect(parseJira(commit)).toEqual(commit);
  });

  test('story found', () => {
    const commit = {
      ...makeCommitFromMsg('Add log'),
      jira: {
        number: ['PLAYA-5052']
      }
    };

    expect(parseJira(makeCommitFromMsg('PLAYA-5052: Add log'))).toEqual(commit);
    expect(parseJira(makeCommitFromMsg('[PLAYA-5052] - Add log'))).toEqual(
      commit
    );
    expect(parseJira(makeCommitFromMsg('[PLAYA-5052] Add log'))).toEqual(
      commit
    );
  });

  test('story found', () => {
    const commit = {
      ...makeCommitFromMsg('Add log'),
      jira: {
        number: ['PLAYA-5052', 'PLAYA-6000']
      }
    };

    expect(
      parseJira(makeCommitFromMsg('PLAYA-5052 PLAYA-6000: Add log'))
    ).toEqual(commit);
    expect(
      parseJira(makeCommitFromMsg('[PLAYA-5052][PLAYA-6000] - Add log'))
    ).toEqual(commit);
    expect(
      parseJira(makeCommitFromMsg('[PLAYA-5052] PLAYA-6000: Add log'))
    ).toEqual(commit);
    expect(
      parseJira(makeCommitFromMsg('PLAYA-5052 [PLAYA-6000] - Add log'))
    ).toEqual(commit);
  });
});

describe('normalizeCommits', () => {
  test('should handle undefined', () => {
    const commits = [makeCommitFromMsg('Filtered', { name: 'pdbf' })];

    expect(normalizeCommits(commits)).toEqual([]);
  });

  test('should do nothing with normal commits', () => {
    const commits = [
      makeCommitFromMsg('First'),
      makeCommitFromMsg('Second'),
      makeCommitFromMsg('Third')
    ];

    expect(normalizeCommits(commits)).toMatchSnapshot();
  });

  test('should use parsing functions to normalize', () => {
    const commits = [
      makeCommitFromMsg('First'),
      makeCommitFromMsg('Second'),
      makeCommitFromMsg('[PLAYA-5052] Add log')
    ];

    expect(normalizeCommits(commits)[2]).toEqual({
      authorEmail: 'adam@dierkens.com',
      authorName: 'Adam Dierkens',
      authors: [
        {
          email: 'adam@dierkens.com',
          name: 'Adam Dierkens'
        }
      ],
      labels: [],
      subject: 'Add log',
      hash: 'foo',
      jira: {
        number: ['PLAYA-5052']
      }
    });
  });
});

const options = {
  owner: 'foobar',
  repo: 'auto-release',
  baseUrl: 'https://github.custom.com/foobar/auto-release',
  jira: 'https://jira.custom.com/browse',
  logger,
  versionLabels: defaultLabels,
  changelogTitles: {
    major: '💥  Breaking Change',
    minor: '🚀  Enhancement',
    patch: '🐛  Bug Fix',
    internal: '🏠  Internal',
    documentation: '📝  Documentation'
  }
};

describe('generateReleaseNotes', () => {
  test('should not create notes for normal commits', () => {
    expect(
      generateReleaseNotes(
        [
          makeCommitFromMsg('First'),
          makeCommitFromMsg('Second'),
          makeCommitFromMsg('Third')
        ],
        dummyLog(),
        options
      )
    ).toBe('');
  });

  test('should create note for PR commits', () => {
    const normalized = normalizeCommits([
      makeCommitFromMsg('First'),
      makeCommitFromMsg('Some Feature (#1234)', { labels: ['minor'] }),
      makeCommitFromMsg('Third')
    ]);

    expect(
      generateReleaseNotes(normalized, dummyLog(), options)
    ).toMatchSnapshot();
  });

  test('should create note for PR commits without labels', () => {
    const normalized = normalizeCommits([
      makeCommitFromMsg('First'),
      makeCommitFromMsg('Some Feature (#1234)'),
      makeCommitFromMsg('Third')
    ]);

    expect(
      generateReleaseNotes(normalized, dummyLog(), options)
    ).toMatchSnapshot();
  });

  test('should create note for jira commits without labels', () => {
    const normalized = normalizeCommits([
      makeCommitFromMsg('First'),
      makeCommitFromMsg('[PLAYA-5052] - Fix P0'),
      makeCommitFromMsg('Third')
    ]);

    expect(
      generateReleaseNotes(normalized, dummyLog(), options)
    ).toMatchSnapshot();
  });

  test('should use username if present', () => {
    const normalized = normalizeCommits([
      makeCommitFromMsg('First'),
      makeCommitFromMsg('Some Feature (#1234)', {
        labels: ['minor'],
        username: 'adam'
      }),
      makeCommitFromMsg('Third')
    ]);

    normalized[1].authors[0].username = 'adam';

    expect(
      generateReleaseNotes(normalized, dummyLog(), options)
    ).toMatchSnapshot();
  });

  test('should create note for JIRA commits', () => {
    const normalized = normalizeCommits([
      makeCommitFromMsg('[PLAYA-5052] - Some Feature (#12345)', {
        labels: ['major'],
        packages: []
      }),
      makeCommitFromMsg('Some Feature (#1234)', { labels: ['internal'] }),
      makeCommitFromMsg('Third', { labels: ['patch'] }),
      makeCommitFromMsg('Fourth')
    ]);

    expect(
      generateReleaseNotes(normalized, dummyLog(), options)
    ).toMatchSnapshot();
  });

  test('should combine pr w/no label and labelled pr', () => {
    const normalized = normalizeCommits([
      makeCommitFromMsg('Some Feature (#1234)'),
      makeCommitFromMsg('Third', { labels: ['patch'] }),
      makeCommitFromMsg('Fourth')
    ]);

    expect(
      generateReleaseNotes(normalized, dummyLog(), options)
    ).toMatchSnapshot();
  });

  test('should create sections for packages', () => {
    const normalized = normalizeCommits([
      makeCommitFromMsg('[PLAYA-5052] - Some Feature (#12345)', {
        labels: ['major'],
        packages: ['foobar/release', 'foobar/party']
      }),
      makeCommitFromMsg('[PLAYA-5052] - Some Feature - Revert (#12345)', {
        labels: ['major'],
        packages: ['foobar/release', 'foobar/party']
      }),
      makeCommitFromMsg('woot (#12343)', {
        labels: ['major']
      }),
      makeCommitFromMsg('Another Feature (#1234)', {
        labels: ['internal'],
        packages: ['foobar/release']
      }),
      makeCommitFromMsg('Third')
    ]);

    expect(
      generateReleaseNotes(normalized, dummyLog(), options)
    ).toMatchSnapshot();
  });

  test("should use only email if author name doesn't exist", () => {
    const commits = normalizeCommits([
      {
        hash: 'foo',
        labels: [],
        authorEmail: 'adam@dierkens.com',
        subject: 'Another Feature (#1234)'
      },
      {
        hash: 'foo',
        labels: [],
        subject: 'One Feature (#1235)'
      }
    ]);

    expect(
      generateReleaseNotes(commits, dummyLog(), options)
    ).toMatchSnapshot();
  });
});
