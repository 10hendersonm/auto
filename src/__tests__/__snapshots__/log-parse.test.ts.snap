// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateReleaseNotes should combine pr w/no label and labelled pr 1`] = `
"#### 🐛  Bug Fix
- Some Feature [#1234](https://github.custom.com/foobar/auto-release/pull/1234) (adam@dierkens.com)
- Third  (adam@dierkens.com)

#### Authors: 1
- Adam Dierkens (adam@dierkens.com)"
`;

exports[`generateReleaseNotes should create note for JIRA commits 1`] = `
"#### 💥  Breaking Change
- [PLAYA-5052](https://jira.custom.com/browse/PLAYA-5052): Some Feature [#12345](https://github.custom.com/foobar/auto-release/pull/12345) (adam@dierkens.com)

#### 🐛  Bug Fix
- Third  (adam@dierkens.com)

#### 🏠  Internal
- Some Feature [#1234](https://github.custom.com/foobar/auto-release/pull/1234) (adam@dierkens.com)

#### Authors: 1
- Adam Dierkens (adam@dierkens.com)"
`;

exports[`generateReleaseNotes should create note for PR commits 1`] = `
"#### 🚀  Enhancement
- Some Feature [#1234](https://github.custom.com/foobar/auto-release/pull/1234) (adam@dierkens.com)

#### Authors: 1
- Adam Dierkens (adam@dierkens.com)"
`;

exports[`generateReleaseNotes should create note for PR commits without labels 1`] = `
"#### 🐛  Bug Fix
- Some Feature [#1234](https://github.custom.com/foobar/auto-release/pull/1234) (adam@dierkens.com)

#### Authors: 1
- Adam Dierkens (adam@dierkens.com)"
`;

exports[`generateReleaseNotes should create note for jira commits without labels 1`] = `
"#### 🐛  Bug Fix
- [PLAYA-5052](https://jira.custom.com/browse/PLAYA-5052): Fix P0  (adam@dierkens.com)

#### Authors: 1
- Adam Dierkens (adam@dierkens.com)"
`;

exports[`generateReleaseNotes should create sections for packages 1`] = `
"#### 💥  Breaking Change
- woot [#12343](https://github.custom.com/foobar/auto-release/pull/12343) (adam@dierkens.com)
- \`foobar/release\`, \`foobar/party\`
  - [PLAYA-5052](https://jira.custom.com/browse/PLAYA-5052): Some Feature [#12345](https://github.custom.com/foobar/auto-release/pull/12345) (adam@dierkens.com)
  - [PLAYA-5052](https://jira.custom.com/browse/PLAYA-5052): Some Feature - Revert [#12345](https://github.custom.com/foobar/auto-release/pull/12345) (adam@dierkens.com)

#### 🏠  Internal
- \`foobar/release\`
  - Another Feature [#1234](https://github.custom.com/foobar/auto-release/pull/1234) (adam@dierkens.com)

#### Authors: 1
- Adam Dierkens (adam@dierkens.com)"
`;

exports[`generateReleaseNotes should use username if present 1`] = `
"#### 🚀  Enhancement
- Some Feature [#1234](https://github.custom.com/foobar/auto-release/pull/1234) ([@adam](https://github.custom.com/adam))

#### Authors: 1
- Adam Dierkens ([@adam](https://github.custom.com/adam))"
`;

exports[`normalizeCommits should do nothing with normal commits 1`] = `
Array [
  Object {
    "author": Object {
      "email": "adam@dierkens.com",
      "name": "Adam Dierkens",
    },
    "authors": Array [
      Object {
        "email": "adam@dierkens.com",
        "name": "Adam Dierkens",
      },
    ],
    "comment": "First",
    "id": "foo",
    "labels": Array [],
    "packages": undefined,
  },
  Object {
    "author": Object {
      "email": "adam@dierkens.com",
      "name": "Adam Dierkens",
    },
    "authors": Array [
      Object {
        "email": "adam@dierkens.com",
        "name": "Adam Dierkens",
      },
    ],
    "comment": "Second",
    "id": "foo",
    "labels": Array [],
    "packages": undefined,
  },
  Object {
    "author": Object {
      "email": "adam@dierkens.com",
      "name": "Adam Dierkens",
    },
    "authors": Array [
      Object {
        "email": "adam@dierkens.com",
        "name": "Adam Dierkens",
      },
    ],
    "comment": "Third",
    "id": "foo",
    "labels": Array [],
    "packages": undefined,
  },
]
`;
